### ISSUE TITLE
rack init failed

### ISSUE THREAD

**Issue by leftwo** [oxidecomputer/omicron#3767]
During tonights dogfood mupdate, we ran into a problem with the final rack init stage.

The following sleds were all going to be part of the cluster, including both 14 and 16
```
Cubby 9 (BRM44220005, bootstrap address fdb0:a840:2504:357::1) 
Cubby 10 (BRM42220009, bootstrap address fdb0:a840:2504:3d4::1) 
Cubby 11 (BRM42220006, bootstrap address fdb0:a840:2504:191::1)
Cubby 12 (BRM42220057, bootstrap address fdb0:a840:2504:393::1)
Cubby 14 (BRM42220051, bootstrap address fdb0:a840:2504:195::1) 
Cubby 16 (BRM42220014, bootstrap address fdb0:a840:2504:352::1)
Cubby 17 (BRM42220017, bootstrap address fdb0:a840:2504:192::1)
Cubby 21 (BRM42220031, bootstrap address fdb0:a840:2504:353::1)
Cubby 23 (BRM42220016, bootstrap address fdb0:a840:2504:395::1)
Cubby 25 (BRM44220010, bootstrap address fdb0:a840:2504:354::1) 
Cubby 27 (BRM42220026, bootstrap address fdb0:a840:2504:351::1)
```

However, the final RSS setup step using wicketd reported `Initialization Failed`

The failed message from wicketd:
```
Status: Initialization Failed 
 Last initialization operation ID: 58b96a33-90ce-4e6e-8cf9-83a98248b387
 Message: Bootstore error: Fsm error: rack init timeout: unacked_peers: {Gimlet { 
 identifier: "BRM42220051", model: "913-0000019", revision: 6 }} 
```

BRM42220051 is sled 14

**Comment by leftwo** [comment]
I'm starting to suspect a restart of sled-agent is the source of the trouble here.

Sled agent logs from 14 and 16 are on catacomb here: /data/staff/core/rack2/omicron-3767

On 14, at 00:50:41, the sled agent restarts:
```
[ Dec 28 00:50:41 Stopping because service disabled. ]
[ Dec 28 00:50:41 Executing stop method (:kill). ]
[ Dec 28 00:51:58 Enabled. ]
[ Dec 28 00:51:58 Executing start method ("ctrun -l child -o noorphan,regent /opt/oxide/sled-agent/sled-agent run /opt/oxide/sled-agent/pkg/config.toml &"). ]
[ Dec 28 00:51:58 Method "start" exited with status 0. ]
note: configured to log to "/dev/stdout"
00:51:58.266Z INFO SledAgent: Starting mg-ddm service
00:51:58.266Z INFO SledAgent: Importing mg-ddm service
    path = /opt/oxide/mg-ddm/pkg/ddm/manifest.xml
00:51:58.448Z INFO SledAgent: Setting mg-ddm interfaces
    interfaces = ("cxgbe0/ll" "cxgbe1/ll")
00:51:58.465Z INFO SledAgent: Enabling mg-ddm service
00:51:58.479Z INFO SledAgent: setting up bootstrap agent server
00:51:58.494Z INFO SledAgent: Ensuring zfs key directory exists
    path = /var/run/oxide/
00:51:59.113Z INFO SledAgent: Sending prefix to ddmd for advertisement
    DdmAdminClient = [::1]:8000
    prefix = Ipv6Prefix { addr: fdb0:a840:2504:195::, len: 64 }
00:51:59.299Z INFO SledAgent: Bootstrap Agent monitoring for hardware
00:51:59.399Z INFO SledAgent (BootstrapAgent): Starting hardware monitor
```
 
I see BootstrapAgent moving along:
```
00:52:01.136Z INFO SledAgent (BootstrapAgent): Initiating connection to new peer: [fdb0:a840:2504:191::1]:12347
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:01.136Z INFO SledAgent (BootstrapAgent): Initiating connection to new peer: [fdb0:a840:2504:192::1]:12347
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:01.137Z INFO SledAgent (BootstrapAgent): Connected to peer
    addr = [fdb0:a840:2504:192::1]:12347
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:01.137Z INFO SledAgent (BootstrapAgent): Connected to peer
    addr = [fdb0:a840:2504:191::1]:12347
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:01.164Z WARN SledAgent (BootstrapAgent): Not configuring DumpSlicePath("/devices/pci@38,0/pci1de,fff9@3,3/pci1344,3100@0/blkdev@w00A0750130084235,0:e") as it appears to contain a dump we cannot yet send to a U.2 debug zvol
00:52:01.338Z INFO SledAgent (BootstrapAgent): Accepted connection from [fdb0:a840:2504:351::1]:53123
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:02.039Z INFO SledAgent (BootstrapAgent): Accepted connection from [fdb0:a840:2504:353::1]:55281
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:02.039Z INFO SledAgent (BootstrapAgent): Accepted connection from [fdb0:a840:2504:352::1]:33182
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:02.061Z INFO SledAgent (BootstrapAgent): Accepted connection from [fdb0:a840:2504:395::1]:47804
```

Then, these:
```
00:52:06.419Z INFO SledAgent (BootstrapAgent): Accepted connection from [fdb0:a840:2504:352::1]:46292
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:06.420Z WARN SledAgent (BootstrapAgent): Connection error: Close
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:06.445Z INFO SledAgent (BootstrapAgent): Accepted connection from [fdb0:a840:2504:1d1::1]:34398
    peer_id = gimlet-BRM42220051-913-0000019-6
00:52:06.445Z WARN SledAgent (BootstrapAgent): Connection error: Close
    peer_id = gimlet-BRM42220051-913-0000019-6
```

Over on 16, we did not get a sled-agent restart.
Things begin here:
```
[ Dec 28 00:00:11 Enabled. ]
[ Dec 28 00:00:45 Executing start method ("ctrun -l child -o noorphan,regent /opt/oxide/sled-agent/sled-agent run /opt/oxide/sled-agent/pkg/config.toml &"). ]
[ Dec 28 00:00:45 Method "start" exited with status 0. ]
note: configured to log to "/dev/stdout"
00:00:47.529Z INFO SledAgent: Starting mg-ddm service
00:00:47.575Z INFO SledAgent: Importing mg-ddm service
    path = /opt/oxide/mg-ddm/pkg/ddm/manifest.xml
00:00:49.241Z INFO SledAgent: Setting mg-ddm interfaces
    interfaces = ("cxgbe0/ll" "cxgbe1/ll")
00:00:49.259Z INFO SledAgent: Enabling mg-ddm service
00:00:49.280Z INFO SledAgent: setting up bootstrap agent server
00:00:49.374Z INFO SledAgent: Ensuring zfs key directory exists
```

Then, when we get close to 00:52...:
```
00:51:57.465Z INFO SledAgent (dropshot (BootstrapAgent)): accepted connection
    local_addr = [fdb0:a840:2504:352::1]:80
    remote_addr = [fdb0:a840:2504:352::2]:32925
00:51:57.465Z INFO SledAgent (dropshot (BootstrapAgent)): request completed
    local_addr = [fdb0:a840:2504:352::1]:80
    method = GET
    remote_addr = [fdb0:a840:2504:352::2]:32925
    req_id = 6ba57792-d01a-4cfe-b56c-bef35678eaf9
    response_code = 200
    uri = /rack-initialize
00:51:57.568Z WARN SledAgent (BootstrapAgent): Fsm error= RackInitTimeout { unacked_peers: {Gimlet { identifier: "BRM42220051", model: "913-0000019", revision: 6 }} }
    peer_id = gimlet-BRM42220014-913-0000019-6
00:51:57.818Z WARN SledAgent (BootstrapAgent): Fsm error= RackInitTimeout { unacked_peers: {Gimlet { identifier: "BRM42220051", model: "913-0000019", revision: 6 }} }
    peer_id = gimlet-BRM42220014-913-0000019-6
00:51:58.069Z WARN SledAgent (BootstrapAgent): Fsm error= RackInitTimeout { unacked_peers: {Gimlet { identifier: "BRM42220051", model: "913-0000019", revision: 6 }} }
    peer_id = gimlet-BRM42220014-913-0000019-6
00:51:58.318Z WARN SledAgent (BootstrapAgent): Fsm error= RackInitTimeout { unacked_peers: {Gimlet { identifier: "BRM42220051", model: "913-0000019", revision: 6 }} }
    peer_id = gimlet-BRM42220014-913-0000019-6
00:51:58.400Z WARN SledAgent (BootstrapAgent): Closing connection: failed to read: Connection reset by peer (os error 131)
    peer_id = gimlet-BRM42220014-913-0000019-6
    remote_peer_id = gimlet-BRM42220051-913-0000019-6
00:51:58.400Z WARN SledAgent (BootstrapAgent): Connection error: Retry
    peer_id = gimlet-BRM42220014-913-0000019-6
    remote_peer_id = gimlet-BRM42220051-913-0000019-6
00:51:58.400Z WARN SledAgent (BootstrapAgent): Missing PeerConnHandle: Stale Disconnected msg
    peer_id = gimlet-BRM42220014-913-0000019-6
    remote_peer_id = gimlet-BRM42220051-913-0000019-6
00:51:58.400Z INFO SledAgent (BootstrapAgent): Connected to peer
    addr = [fdb0:a840:2504:195::1]:12347
    peer_id = gimlet-BRM42220014-913-0000019-6
00:51:58.425Z INFO SledAgent (dropshot (BootstrapAgent)): accepted connection
    local_addr = [fdb0:a840:2504:352::1]:80
    remote_addr = [fdb0:a840:2504:352::2]:55990
00:51:58.425Z INFO SledAgent (dropshot (BootstrapAgent)): request completed
    local_addr = [fdb0:a840:2504:352::1]:80
    method = GET
    remote_addr = [fdb0:a840:2504:352::2]:55990
    req_id = 3a2796cf-54c6-4d72-b8a7-44aacb30e3d8
    response_code = 200
    uri = /rack-initialize
00:51:58.453Z WARN SledAgent (BootstrapAgent): Missing PeerConnHandle; Stale ConnectedInitiator msg
    addr = [fdb0:a840:2504:195::1]:12347
    peer_id = gimlet-BRM42220014-913-0000019-6
    remote_peer_id = gimlet-BRM42220051-913-0000019-6
```

Which, makes me suspect that the reset on 14 may have started the problems that are
preventing the rack init from going through.

**Comment by andrewjstone** [comment]
This is believed to be fixed by #3771 as we have successfully run bootstore rack init twice on dogfood since that fix was added. However, I don't like how unclear some of the code for handling connections is and I think it deserves some cleanup and tests. I'm therefore keeping this open and assigning it to myself with the hope that I can work on that tomorrow.

**Comment by andrewjstone** [comment]
Afaik, we haven't seen this again after #3771. I don't think it's worth focusing on further cleanup for lrtq. I'd rather start on the real tq design and impl.
